Sources/SwiftFusion/Inference/PPCATrackingFactor.swift

```diff
@@ -0,0 +1,119 @@
+import PenguinStructures
+import TensorFlow
+
+/// A factor over a target's pose and appearance in an image.
+public struct PPCATrackingFactor: LinearizableFactor2 {
+  /// The first adjacent variable, the pose of the target in the image.
+  public typealias V0 = Pose2
+
+  /// The second adjacent variable, the PPCA latent code for the appearance of the target.
+  public typealias V1 = Vector5
+
+  /// A region cropped from the image.
+  public typealias Patch = Tensor28x62x1
+
+  /// The IDs of the variables adjacent to this factor.
+  public let edges: Variables.Indices
+
+  /// The image containing the target.
+  public let measurement: Tensor<Double>
+
+  /// The PPCA appearance model weight matrix.
+  ///
+  /// The shape is `Patch.shape + [V1.dimension]`.
+  public var W: Tensor<Double>
+
+  /// The PPCA appearance model mean.
+  public var mu: Patch
+
+  /// Creates an instance.
+  ///
+  /// - Requires `W.shape == Patch.shape + [V1.dimension]`.
+  public init(_ poseId: TypedID<Pose2>, _ latentId: TypedID<Vector5>, measurement: Tensor<Double>, W: Tensor<Double>, mu: Patch) {
+    precondition(W.shape == Patch.shape + [V1.dimension])
+    self.edges = Tuple2(poseId, latentId)
+    self.measurement = measurement
+    self.W = W
+    self.mu = mu
+  }
+
+  /// Returns the difference between the PPCA generated `Patch` and the `Patch` cropped from
+  /// `measurement`.
+  @differentiable
+  public func errorVector(_ pose: Pose2, _ latent: Vector5) -> Patch.TangentVector {
+    let bbox = OrientedBoundingBox(center: pose, rows: Patch.shape[0], cols: Patch.shape[1])
+    let generatedAppearance =
+      mu + Patch(matmul(W, latent.flatTensor.expandingShape(at: 1)).squeezingShape(at: 2))
+    return generatedAppearance - Patch(measurement.patch(at: bbox))
+  }
+
+  /// Returns a linear approximation to `self` at `x`.
+  public func linearized(at x: Variables) -> LinearizedPPCATrackingFactor {
+    let pose = x.head
+    let latent = x.tail.head
+
+    let bbox = OrientedBoundingBox(center: pose, rows: Patch.shape[0], cols: Patch.shape[1])
+    let generatedAppearance =
+      mu + Patch(matmul(W, latent.flatTensor.expandingShape(at: 1)).squeezingShape(at: 2))
```

Comment:
When linearizing a factor graph, we only run the `linearized` method because it calculates everything required for the linearization, including the error vector. So it's not duplicate work.

This comment did make me notice that this is duplicate code, so I'll factor it out into another method.